pyrokinetics/gk_code/GKInputReaderGS2.py

```diff
@@ -0,0 +1,244 @@
+import numpy as np
+import f90nml
+from cleverdict import CleverDict
+from copy import copy
+
+from ..typing import PathLike
+from ..constants import pi, sqrt2
+from ..local_species import LocalSpecies
+from ..local_geometry import LocalGeometry, LocalGeometryMiller, default_miller_inputs
+from ..numerics import Numerics
+from .GKInputReader import GKInputReader
+from .gs2_utils import pyro_gs2_miller, pyro_gs2_species, gs2_is_nonlinear
+
+
+class GKInputReaderGS2(GKInputReader):
+    """
+    Class that can read GS2 input files, and produce
+    Numerics, LocalSpecies, and LocalGeometry objects
+    """
+
+    code_name = "GS2"
+
+    def read(self, filename: PathLike) -> f90nml.Namelist:
+        """
+        Reads GS2 input file into a dictionary
+        """
+        super().read(filename)
+        # TODO These lines correct a possible mistake in the input file. Why is
+        #      this mistake corrected in particular? Should these corrections be
+        #      expanded?
+        if self.is_nonlinear() and "wstar_units" in self.data["knobs"]:
+            self.data["knobs"].pop("wstar_units")
+        return self.data
+
+    def verify(self, filename: PathLike):
+        """
+        Ensure this file is a valid gs2 input file"
+        """
+        data = f90nml.read(filename).todict()
+        expected_keys = ["kt_grids_knobs", "theta_grid_knobs", "theta_grid_eik_knobs"]
+        if not np.all(np.isin(expected_keys, list(data))):
+            raise ValueError(f"Expected GS2 file, received {filename}")
```

Comment:
An issue here is that `verify` must be able to uniquely identify a file type in some way, so testing for a `.in` extension could return a false positive if any other code uses the same convention. This isn't too serious, as the user can override file type inference by passing `gk_input_type="GS2"`, but it could lead to some confusion if a user tries inputting a GENE file and Pyrokinetics thinks it's looking at a GS2 file. Are we aware of any other codes that use `.in`?

Reply:
I don't know of any codes which use that, but wouldn't like to say that there aren't any! Could it work by eliminating possibilities, rather than trying to guess the right option? Each code type checks essential conditions of the input file and then if there's more than one that doesn't complain we raise an error saying it is ambiguous so the user has to specify?

~~~~
possible_codes = [
       x for x in GKInputReaders if x.compatible_input(input)
]

if len(possible_codes) > 1:
   raise ....
elif len(possible_codes) < 1:
   raise .... 
else:
   reader = possible_codes[0]
~~~~

Reply:
In the first branch we could even go  over the possible codes and ask each to look for optional things which may give a hint. For example GS2 could check if any of its namelists are found, if so it can be quite confident that it is probably a GS2 input.

Reply:
Even if nothing currently has a ```.in``` file (which `STELLA` might) I wouldn't want to make sure we can integrate a new code that could have this suffix. This is a tricky problem overall. Bar running the codes themselves it will be difficult to 100% guarantee the correct code is being read. This may just be a few edge cases though

Reply:
Yes, I think this should be a "best effort guess", and the user can always override that if it's wrong. In the factory maybe we want some output like "gk_code not specified, guessing filetype as GS2; set `gk_code` to override"

I like David's suggestion of turning this around so that the function says "this is consistent with a GS2 input file"

Reply:
Going forward, I think you're right that a 'best effort guess' would be a better system. However, if you try to use a blank input file with Pyrokinetics and specify `gk_input_type="GS2"`, it currently doesn't know what to do with it, as it requires an input file to include, at a minimum: `kt_grids_knobs`, `theta_grid_knobs`, `theta_grid_eik_knobs` (plus some others I should add to the `expected_keys` list).

For now, I think it would be better for `verify` to test "can Pyrokinetics parse this as a GS2 file", rather than "is this a valid GS2 input". In future, I think it would be better if Pyrokinetics could fill in GS2 defaults automatically so that a blank input file would work as expected, but that's a separate PR.